### PR TITLE
Add description to task.onlyIf specs

### LINKED ISSUE #7541: Add ability to specify "name" for task.onlyIf actions (and possibly others like upToDateWhen)
`doFirst` and `doLast` offer the ability to specify a description/name for the action. It would be 

### Expected Behavior

```
Skipping task ':MyTask' as task onlyIf "only when executing in CI environment" is false.
```

### Current Behavior

```
Skipping task ':MyTask' as task onlyIf is false.
```

### Context

It is difficult to understand an `onlyIf`, especially if it is a composed `onlyIf` of multiple actions. It would be useful if the logs and other pieces told you why it was not executed.

### PR DESCRIPTION
**Pull request by urdak**
Fixes #7541

### Context

Adds overloads to `Task::onlyIf` which include a description for the spec. This description is logged when the `onlyIf` condition is not met.

The changes to `AndSpec` were made to avoid duplicating the same logic for `onlyIf`. Access to the underlying specs array in [`CompositeSpec` is package private](https://github.com/gradle/gradle/blob/16fbec5c822c9954894e5d8acd95a1725f2c3bf2/subprojects/base-services/src/main/java/org/gradle/api/specs/CompositeSpec.java#L71), so efficient iteration (which seems to be required for large projects) can only be done in the core API scope.

`DescribingAndSpec` uses the already existing [`SelfDescribingSpec`](https://github.com/gradle/gradle/blob/16fbec5c822c9954894e5d8acd95a1725f2c3bf2/subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SelfDescribingSpec.java) that is currently used for [`TaskOutputs::cacheIf`](https://github.com/gradle/gradle/blob/16fbec5c822c9954894e5d8acd95a1725f2c3bf2/subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskOutputs.java#L96)/ [`TaskOutputs::doNotCacheIf`](https://github.com/gradle/gradle/blob/16fbec5c822c9954894e5d8acd95a1725f2c3bf2/subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskOutputs.java#L112) descriptions.

If the proposed implementation is acceptable, the same functionality can be easily added to [`TaskOutputs::upToDateWhen`](https://github.com/gradle/gradle/blob/16fbec5c822c9954894e5d8acd95a1725f2c3bf2/subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskOutputs.java#L65).
The `cacheIf` and `doNotCacheIf` methods can also be re-implemented using this approach to [cleanup some usages](https://github.com/gradle/gradle/blob/16fbec5c822c9954894e5d8acd95a1725f2c3bf2/subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/DefaultTaskCacheabilityResolver.java#L90-L106).

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [x] Recognize contributor in release notes

### CODE REVIEW

**Review by facewindu — APPROVED** [review]

**Review by eskatos — APPROVED** [review]
:100: 

I think this is release-notes-worthy and should be mentioned.

**Review by alllex — APPROVED** [review]

File: subprojects/docs/src/docs/userguide/dep-man/06-publishing/publishing_customization.adoc [hunk 1]
```diff
@@ -140,7 +140,7 @@ include::sample[dir="snippets/maven-publish/conditional-publishing/kotlin",files
 
 The publishing plugins will create tasks that allow you to publish either of the publications to either repository. They also attach those tasks to the `publish` aggregate task. But let's say you want to restrict the binary-only publication to the external repository and the binary-with-sources publication to the internal one. To do that, you need to make the publishing _conditional_.
 
-Gradle allows you to skip any task you want based on a condition via the link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:onlyIf(org.gradle.api.specs.Spec)[Task.onlyIf(org.gradle.api.specs.Spec)] method. The following sample demonstrates how to implement the constraints we just mentioned:
+Gradle allows you to skip any task you want based on a condition via the link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:onlyIf(java.lang.String,org.gradle.api.specs.Spec)[Task.onlyIf(org.gradle.api.specs.Spec)] method. The following sample demonstrates how to implement the constraints we just mentioned:
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/dep-man/06-publishing/publishing_customization.adoc:143]: The link has been changed but not the text.
💄 Let's also move to one sentence per line.

**Review by ljacomet — CHANGES_REQUESTED** [review]
I agree with @eskatos that this should be added to the release notes, but it can be done in a follow up PR.

See below for a minor missed doc update:

File: subprojects/docs/src/docs/userguide/dep-man/06-publishing/publishing_customization.adoc (hunk 1 above)
Comment (h0tk3y) [reply subprojects/docs/src/docs/userguide/dep-man/06-publishing/publishing_customization.adoc]: Thanks, fixed it.

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/TaskOnlyIfReasonIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ */
+
+package org.gradle.api.internal.tasks
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.api.internal.tasks.execution.ExecuteTaskBuildOperationType
+
+class TaskOnlyIfReasonIntegrationTest extends AbstractIntegrationSpec {
+    def 'task only if with description prints value'() {
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/TaskOnlyIfReasonIntegrationTest.groovy]: Fixed it. Thanks!

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Comment by h0tk3y** [comment]
@bot-gradle clean

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

File: subprojects/enterprise-operations/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteTaskBuildOperationType.java [hunk 3]
```diff
@@ -62,6 +62,16 @@ public interface Result {
...
 
+        /**
+         * The detailed reason why the task was skipped, if provided by the project configuration.
+         *
+         * Expected to be {@link org.gradle.api.tasks.TaskState#getSkipReasonMessage()}.
+         *
+         * @since 7.6
+         */
+        @Nullable
+        String getSkipReasonMessage();
```
Comment (ldaley) [inline subprojects/enterprise-operations/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteTaskBuildOperationType.java:73]: Is there anything that we can codify about the presence of this value? e.g. is it only present with #getSkipMessage() has a certain value?

Comment (h0tk3y) [reply subprojects/enterprise-operations/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteTaskBuildOperationType.java:73]: Yes, at least for now, the skip reason message is set if and only if the outcome (skip message) is `SKIPPED`. If the task is skipped but the user has not provided a reason message (e.g. `onlyIf { false }`), then there goes a generated message like `'Task satisfies onlyIf spec' not satisfied`.

Comment (ldaley) [reply subprojects/enterprise-operations/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteTaskBuildOperationType.java:73]: @h0tk3y Can you please include this information in the Javadoc here.

**Pull request closed without merging**